src/librustc_mir/borrow_check/nll/universal_regions.rs

```diff
@@ -490,6 +490,9 @@ impl<'cx, 'gcx, 'tcx> UniversalRegionsBuilder<'cx, 'gcx, 'tcx> {
             &bound_inputs_and_output,
             &mut indices,
         );
+
+        self.infcx.replace_late_bound_regions_with_nll_infer_vars(self.mir_def_id, &mut indices);
```

Comment:
As I noted on Zulip, in the case of a closure, this code needs to execute earlier -- that is, the late-bound regions on the closure's creator are not *local* free regions, so they must be created up above, before the `first_local_index` variable is assigned.